seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/canal/CanalJsonDeserializationSchema.java

```diff
@@ -113,93 +114,93 @@ public SeaTunnelDataType<SeaTunnelRow> getProducedType() {
-                assert after != null;
-                after.setRowKind(RowKind.UPDATE_AFTER);
-                out.collect(before);
-                out.collect(after);
+                throw new IllegalStateException(
+                        format("Null data value '%s' Cannot send downstream", jsonNode));
             }
-        } else if (OP_DELETE.equals(type)) {
-            for (int i = 0; i < dataNode.size(); i++) {
-                SeaTunnelRow row = convertJsonNode(dataNode.get(i));
-                assert row != null;
-                row.setRowKind(RowKind.DELETE);
-                out.collect(row);
+            if (OP_INSERT.equals(type)) {
+                for (int i = 0; i < dataNode.size(); i++) {
+                    SeaTunnelRow row = convertJsonNode(dataNode.get(i));
+                    out.collect(row);
+                }
+            } else if (OP_UPDATE.equals(type)) {
+                final ArrayNode oldNode = (ArrayNode) jsonNode.get(FIELD_OLD);
+                for (int i = 0; i < dataNode.size(); i++) {
+                    SeaTunnelRow after = convertJsonNode(dataNode.get(i));
+                    SeaTunnelRow before = convertJsonNode(oldNode.get(i));
+                    for (int f = 0; f < fieldCount; f++) {
+                        if (before.isNullAt(f) && oldNode.findValue(fieldNames[f]) == null) {
+                            // fields in "old" (before) means the fields are changed
+                            // fields not in "old" (before) means the fields are not changed
+                            // so we just copy the not changed fields into before
+                            before.setField(f, after.getField(f));
+                        }
+                    }
+                    before.setRowKind(RowKind.UPDATE_BEFORE);
+                    after.setRowKind(RowKind.UPDATE_AFTER);
+                    out.collect(before);
+                    out.collect(after);
+                }
+            } else if (OP_DELETE.equals(type)) {
+                for (int i = 0; i < dataNode.size(); i++) {
+                    SeaTunnelRow row = convertJsonNode(dataNode.get(i));
+                    row.setRowKind(RowKind.DELETE);
+                    out.collect(row);
+                }
+            } else {
+                throw new IllegalStateException(format("Unknown operation type '%s'.", type));
             }
-        } else {
+        } catch (Throwable t) {
             if (!ignoreParseErrors) {
-                throw new SeaTunnelJsonFormatException(
-                        CommonErrorCodeDeprecated.UNSUPPORTED_DATA_TYPE,
-                        format(
-                                "Unknown \"type\" value \"%s\". The Canal JSON message is '%s'",
-                                type, jsonNode.asText()));
+                Throwable cause = CommonError.jsonOperationError(FORMAT, jsonNode.toString(), t);
+                throw new IOException(cause);
```

Comment:
```suggestion
            throw CommonError.jsonOperationError(FORMAT, jsonNode.toString(), t);
```

Reply:
Just throw `SeaTunnelRuntimeException`

Reply:
Why not follow method signatures?

Reply:
We should ensure that our error messages are clear and effective, and there is no need to create an additional layer of packaging. At the same time, in most cases, the outer layer should not catch SeaTunnelRuntimeException. If it is an IOException, it may be caught.